analyze-network-with-subnetwork-trace/src/main/java/com/esri/arcgismaps/sample/analyzenetworkwithsubnetworktrace/MainActivity.kt

```diff
@@ -0,0 +1,512 @@
+        }
+
+        // get the currently selected attribute
+        sourcesList?.get(sourcePosition)?.let { sourceAttribute ->
+            // if the other value is a coded value domain
+            val otherValue = if (sourceAttribute.domain is CodedValueDomain) {
+                codedValuesList?.get(valuePosition)?.code?.let {
+                    convertToDataType(it, sourceAttribute.dataType)
+                }
+            } else {
+                convertToDataType(valuesEditText.text.toString(), sourceAttribute.dataType)
+            }
+
+            if (otherValue.toString().contains("Error") || otherValue == null) {
+                return showError(otherValue.toString())
+            }
+
+            // get the currently selected attribute operator>
+            operatorsList?.get(operatorPosition)?.let { comparisonOperator ->
+                // NOTE: You may also create a UtilityNetworkAttributeComparison
+                // with another NetworkAttribute
+                var expression: UtilityTraceConditionalExpression =
+                    UtilityNetworkAttributeComparison(
+                        sourceAttribute,
+                        comparisonOperator,
+                        otherValue
+                    )
+                (traceConfiguration.traversability?.barriers as? UtilityTraceConditionalExpression)?.let { otherExpression ->
+                    // NOTE: You may also combine expressions with UtilityTraceAndCondition
+                    expression = UtilityTraceOrCondition(otherExpression, expression)
+                }
+                traceConfiguration.traversability?.barriers = expression
+                expressionTextView.text = expressionToString(expression)
+            }
+        }
+    }
+
+    /**
+     * Show the UI of the given [id] and hide the others which share the same space.
+     */
+    private fun setVisible(id: Int) {
+        when (id) {
+            valuesBackgroundView.id -> {
+                valuesBackgroundView.visibility = View.VISIBLE
+                valueBooleanButton.visibility = View.GONE
+                valuesEditText.visibility = View.GONE
+            }
+            valuesEditText.id -> {
+                valuesEditText.visibility = View.VISIBLE
+                valueBooleanButton.visibility = View.GONE
+                valuesBackgroundView.visibility = View.GONE
+            }
+            valueBooleanButton.id -> {
+                valueBooleanButton.visibility = View.VISIBLE
+                valuesBackgroundView.visibility = View.GONE
+                valuesEditText.visibility = View.GONE
+            }
+        }
+    }
+
+    /**
+     * Run the network trace with the parameters and display the result in an alert dialog
+     * when the [traceButton] is clicked.
+     */
+    fun trace(traceButton: View) {
+        if (utilityNetwork.loadStatus.value != LoadStatus.Loaded) {
+            return showError("Utility network is not loaded")
```

Comment:
A resource String?

Reply:
Will leave it as it is, the in-line message is usually helpful in understanding the cause of the error while reading through the code. Samples usually use String resources when it's a resource for a `TextView` or to reuse repeated resources.